### ISSUE TITLE
Create rule XXX: "filter" should be called before "map" in a chain call

### ISSUE THREAD

**Issue by yassin-kammoun-sonarsource** [SonarSource/SonarJS#4411]
### Noncompliant

```js
const numbers = [1, 2, 3, 4, 5];

const result = numbers
  .map(number => number * 2)
  .filter(number => number % 2 === 0); // Noncompliant

console.log(result);
```

### Compliant

```js
const numbers = [1, 2, 3, 4, 5];

const result = numbers
  .filter(number => number % 2 === 0)
  .map(number => number * 2);

console.log(result);
```

**Comment by erichirlemann** [comment]
Interesting idea, but the filtering criteria may happen on data issued by the `map` transformation. 
I mean this would be a false positive:
```typescript
const coloredBalls <Cat,Ball> = new Map();
...
const redBalls = cats 
  .map(cat => coloredBalls.get(cat));
  .filter(ball => ball.color === 'red')
```

But the idea may eventually be used on checking the order of `slice` and `map` instead:

Bad code:
```typescript
const tenBalls = cats 
  .map(cat => coloredBalls.get(cat));
  .slice(0,10); 
```

Good code:
```typescript
const tenBalls = cats 
  .slice(0,10)
  .map(cat => coloredBalls.get(cat));
```

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-69
